### PR TITLE
Escape RST substitution only when needed

### PR DESCRIPTION
**Pull request by davfsa**
I was running into some issues with a random `\` being added at the beginning of my draft changelog. I am not using the default RST parser, instead I am making use of a Markdown one, which made the character visible instead of rendering as an empty character.

Hopefully adding it only when needed is a valid solution to this issue! Let me know if there is anything I am missing from the PR.

Would also like to kindly ask for a release when this PR is merged, as the project seems stable enough and doesnt receive any consistent releases :)

### CODE REVIEW

File: src/sphinxcontrib/towncrier/__init__.py [hunk 1]
```diff
@@ -53,18 +53,20 @@ def _get_changelog_draft_entries(
...
-        # NOTE: The escaped space sequence (`\ `) is necessary to address
-        # NOTE: a corner case when the towncrier config has something like
-        # NOTE: `v{version}` in the title format **and** the directive target
-        # NOTE: argument starts with a substitution like `|release|`. And so
-        # NOTE: when combined, they'd produce `v|release|` causing RST to not
-        # NOTE: substitute the `|release|` part. But adding an escaped space
-        # NOTE: solves this: that escaped space renders as an empty string and
-        # NOTE: the substitution gets processed properly so the result would
-        # NOTE: be something like `v1.0` as expected.
+        target_version,  # version value to be used in the RST title
```
Comment (webknjaz) [inline src/sphinxcontrib/towncrier/__init__.py:69]:
```suggestion
        target_version,  # version value to be used in the RST title
```
(trailing whitespaces)

**Comment by webknjaz** [comment]
> Would also like to kindly ask for a release when this PR is merged, as the project seems stable enough and doesnt receive any consistent releases :)

It's quite stable, yes. I've been holding off dropping the alpha marker, though, until there's decent testing implemented...

**Comment by webknjaz** [comment]
> instead I am making use of a Markdown one, which made the character visible instead of rendering as an empty character.

Oh, that's interesting. Are you talking about MyST? I don't think I've faced problems with it. Do you have an example project where this is happening?

**Comment by webknjaz** [comment]
pre-commit.ci run

**Comment by davfsa** [comment]
> Oh, that's interesting. Are you talking about MyST? I don't think I've faced problems with it. Do you have an example project where this is happening?

Yep, im using MyST. This is what my changelog.md file looks like

````
# Changelog

```{attention}
Mayor and minor releases also include the changes specified in prior development releases.
```

```{towncrier-draft-entries} Unreleased changes
``` 

```{include} ../../CHANGELOG.md
``` 
````

**Comment by webknjaz** [comment]
@davfsa oh, now I see why I've never seen this issue: https://github.com/ansible/ansible-navigator/blob/main/docs/changelog.md?plain=1#L22 — I just didn't set the version title value.

File: src/sphinxcontrib/towncrier/__init__.py [hunk 2]
```diff
@@ -53,18 +53,20 @@ def _get_changelog_draft_entries(
         config_path: str = None,
 ) -> str:
     """Retrieve the unreleased changelog entries from Towncrier."""
+    if target_version.startswith('|') and target_version.count('|') > 1:
```
Comment (webknjaz) [inline src/sphinxcontrib/towncrier/__init__.py:56]: @davfsa I suggest moving this escaping into a separate pure function so that we could write a test for it specifically. This bit is obscure and easy enough to break unintentionally so I'm now convinced that for this PR to be truly valuable, there must be a regression test bundled with it. Sounds good?

Comment (davfsa) [reply src/sphinxcontrib/towncrier/__init__.py:56]: Sounds great, will do when I get around to! Hopefully later today or early tomorrow

Comment (webknjaz) [reply src/sphinxcontrib/towncrier/__init__.py:56]: Thanks! No rush tho :)

Comment (davfsa) [reply src/sphinxcontrib/towncrier/__init__.py]: Done!

File: .flake8 [hunk 3]
```diff
@@ -60,13 +60,15 @@ per-file-ignores =
   # WPS412 logic of an extension is in __init__.py file;
   # FIXME: WPS201 too many imports
   # FIXME: WPS402 too many `noqa`s
-  src/sphinxcontrib/towncrier/__init__.py: E800, WPS201, WPS402, WPS412
+  # FIXME: WPS202 too many module members
```
Comment (webknjaz) [inline .flake8:63]: If we just ignore every linting rule that annoys us, this would make using the linters pointless.
There's no good reason for ignoring this. If the rule is hit, this calls for splitting a module into two or more. Just like it was done with `_towncrier`.
Besides, it's configurable via `max-module-members` setting that defaults to `7` currently: https://wemake-python-styleguide.readthedocs.io/en/latest/pages/usage/violations/complexity.html#wemake_python_styleguide.violations.complexity.TooManyModuleMembersViolation.

File: src/sphinxcontrib/towncrier/__init__.py [hunk 4]
```diff
@@ -45,6 +45,22 @@
...
+    if version.startswith('|') and version.count('|') > 1:
+        # A corner case exists when the towncrier config has something like
+        # `v{version}` in the title format **and** the directive target
+        # argument starts with a substitution like `|release|`. And so
+        # when combined, they produce a v|release|` causing RST to not
+        # substitute the `|release|` part. But adding an escaped space
+        # solves this: that escaped space renders as an empty string and
+        # the substitution gets processed properly so the result would
+        # be something like `v1.0` as expected.
+        version = rf'\ {version}'
```
Comment (webknjaz) [inline src/sphinxcontrib/towncrier/__init__.py:59]: Could just
```suggestion
        return rf'\ {version}'
```

File: src/sphinxcontrib/towncrier/__init__.py [hunk 5]
```diff
@@ -45,6 +45,22 @@
 logger = logging.getLogger(__name__)
 
 
+def _escape_version(version: str) -> str:
```
Comment (webknjaz) [inline src/sphinxcontrib/towncrier/__init__.py:48]: This function name needs to be descriptive. Something like:
```suggestion
def _escape_project_version_for_markup_source(version: str) -> str:
```

File: src/sphinxcontrib/towncrier/__init__.py [hunk 6]
```diff
@@ -45,6 +45,22 @@
...
+    """Escape version as needed to cover edge cases."""
```
Comment (webknjaz) [inline src/sphinxcontrib/towncrier/__init__.py:49]: The docstring should be clear about what the function does. "as needed to cover edge cases" is obscure and would be confusing to read for anybody interactively inspecting the functions in REPL or autogenerated docs. Besides, it doesn't actually escape the version, it just escapes the beginning of a possible RST substitution conditionally. In fact, there is no guarantee there would be any mentions of version present in that string.

File: tests/extension_test.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+
+def test_escape_version():
+    """Test that escaping the version covers edge case.
+
+    RST substitution as the first item should be escaped.
+    """
+    assert _escape_version('|release|') == r'\ |release|'
+    assert _escape_version('|release') == '|release'
+    assert _escape_version('v|release|') == 'v|release|'
```
Comment (webknjaz) [inline tests/extension_test.py:12]: These are 3 separate tests, they shouldn't appear as a single test in the pytest summary. Instead, use `@pytest.mark.parametrize` to declare different params for the same test function.

File: tests/extension_test.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
+"""Extension tests."""
```
Comment (webknjaz) [inline tests/extension_test.py:1]: Why extension? Did you mean escaping?

File: .flake8 [hunk 9]
```diff
@@ -60,13 +60,15 @@ per-file-ignores =
...
+  src/sphinxcontrib/towncrier/__init__.py: E800, WPS201, WPS402, WPS412, WPS202
 
   # There are multiple `assert`s (S101)
   # and subprocesses (import – S404; call – S603) in tests;
   # also, using fixtures looks like shadowing the outer scope (WPS442);
+  # furthermore, we should be able to import and test private attributes (WPS450)
```
Comment (webknjaz) [inline .flake8:69]: For now, don't add this. Instead, ignore inline in a specific place. Adding global-ish ignores should be the last resort or should have a really good reason.

Besides, if you just follow the example of the `_towncrier` module, maybe there would be no need in ignoring anything at all.

**Review by webknjaz — CHANGES_REQUESTED** [review]
There's a number of things to be fixed before this is merge-worthy. The biggest concern being disabling linting. Also, there's maintainability improvements needing to be done. I've added several comments below, with more explanation as to why.

File: src/sphinxcontrib/towncrier/__init__.py (hunk 4 above)
Comment (davfsa) [reply src/sphinxcontrib/towncrier/__init__.py:59]: Could, the reason I didn't do it is to allow for just adding more rules of things that might need to be escaped. Will change tho

Comment (webknjaz) [reply src/sphinxcontrib/towncrier/__init__.py:59]: That's sounds like a premature optimization. FWIW, let's optimize for "here and now". It's not like this can't be change when the need be. Right now, having early exit instead of a variable mutation seems like an easier to follow control flow.

File: .flake8 (hunk 3 above)
Comment (davfsa) [reply .flake8:63]: I am uncertain how you would rather them be split. I personally dont see the need for an extra module just for this one function which will only really be used under this specific context. `_towncrier` seem to contain a function that aids with getting the configuration, but other functions like `_lookup_towncrier_fragments` and `_get_changelog_draft_entries` are placed in the `__init__`. What would be the prefered criteria for when to split/how to place them?

File: .flake8 (hunk 9 above)
Comment (webknjaz) [reply .flake8:77]: I gave it some thought, and maybe it's reasonable to have this ignore just for the tests directory. I'll cherry-pick it separately, though. Because it doesn't seem to be something directly related to the PR purpose. It's better if it remains atomic.

File: .flake8 (hunk 3 above)
Comment (webknjaz) [reply .flake8:71]: I'm not sure about the criteria but I'd be fine with putting a single function into a separate module.
There's another (hidden?) problem here — it is an antipattern to have the logic in an `__init__.py` module because this sort of thing causes side-effects during import-time. And so it's usually better to have an entry point in some `.api` submodule.
Obviously, it's not to be solved in this PR but it'd be good to do it at some point in the future.

FWIW I'll try to come up with an intermediate solution myself for now.

**Comment by webknjaz** [comment]
This has been released in v0.3.0a0: https://pypi.org/project/sphinxcontrib-towncrier/0.3.0a0/.
